pkg/store/storepb/custom.go

```diff
@@ -517,3 +517,27 @@ func (c *SeriesStatsCounter) Count(series *Series) {
 		}
 	}
 }
+
+func (m *SeriesRequest) ToPromQL() string {
+	return m.QueryHints.toPromQL(m.Matchers)
+}
+
+// IsSafeToExecute returns true if the function or aggregation from the query hint
+// can be safely executed by the underlying Prometheus instance without affecting the
+// result of the query.
+func (m *QueryHints) IsSafeToExecute() bool {
+	distributiveOperations := []string{
+		"max",
+		"max_over_time",
+		"min",
+		"min_over_time",
+		"group",
```

Comment:
Why not `count()` btw.? I guess some others would also work, but I guess you're just starting with a few...

Reply:
Because then PromQL engine goes ahead and calculates count() again on the already correct results i.e. pushed down data thus corrupting it. I have suggested [something here](https://github.com/prometheus/prometheus/issues/10101) but it got no attention so I haven't pursued it further.

Reply:
Ah right, I had assumed there was already some facility for changing the outer operation (in the `count` case, to `sum`), but that's not required for min/max/group.

The comment on https://github.com/prometheus/prometheus/issues/10101 is interesting though. Indeed, if the same series can exist on two leafs (I think this is only the case for HA deduplication in Thanos, right?), then pushing `rate()` or even `count()` down in a correct way would be impossible (if you do the count across two HA replicas and you get 10 on each, you don't know if you really counted the same 10 series both times or counted different ones, or some overlapping sets).